### PR TITLE
add debug logs for blob uploads & downloads

### PR DESCRIPTION
**Pull request by thundergolfer**
## Describe your changes

Motivated by a couple of recent customer calls where we debugging 'maybe slow, maybe stuck' Modal Apps. I brought this up in late September and I'm partly actioning it here https://modal-com.slack.com/archives/C056CGAANRM/p1727275139669729. 

In the customer call today their script was actually stuck on **`print`** and not the download of the large output. We could have confirmed this earlier with more verbose debug logging.

### CODE REVIEW

File: modal/_utils/blob_utils.py [hunk 1]
```diff
@@ -317,11 +324,16 @@ async def _download_from_url(download_url: str) -> bytes:
...
+    """Convenience function reading all of the downloaded file into memory."""
+    logger.debug(f"Downloading large blob {blob_id}")
+    t0 = time.time()
     req = api_pb2.BlobGetRequest(blob_id=blob_id)
     resp = await retry_transient_errors(stub.BlobGet, req)
-
-    return await _download_from_url(resp.download_url)
+    data = await _download_from_url(resp.download_url)
+    size_mib = len(data) / 1024 / 1024
+    throughput_mib_s = (len(data) / 1024 / 1024) / (time.time() - t0)
```
Comment (mwaskom) [inline modal/_utils/blob_utils.py:334]: Nit
```suggestion
    throughput_mib_s = size_mib / (time.time() - t0)

```

File: modal/_utils/blob_utils.py [hunk 2]
```diff
@@ -317,11 +324,16 @@ async def _download_from_url(download_url: str) -> bytes:
 
 
 async def blob_download(blob_id: str, stub: ModalClientModal) -> bytes:
-    # convenience function reading all of the downloaded file into memory
+    """Convenience function reading all of the downloaded file into memory."""
```
Comment (mwaskom) [inline modal/_utils/blob_utils.py:327]: Missing a word? ("Convenience function *for* ...)?

File: modal/_utils/blob_utils.py [hunk 3]
```diff
@@ -288,11 +289,17 @@ async def _blob_upload(
...
+    logger.debug(f"Uploading large blob of size {len(payload) / 1024 / 1024:.2f} MiB")
+    t0 = time.time()
     if isinstance(payload, str):
         logger.warning("Blob uploading string, not bytes - auto-encoding as utf8")
         payload = payload.encode("utf8")
     upload_hashes = get_upload_hashes(payload)
-    return await _blob_upload(upload_hashes, payload, stub)
+    blob_id = await _blob_upload(upload_hashes, payload, stub)
+    size_mib = len(payload) / 1024 / 1024
+    throughput_mib_s = (len(payload) / 1024 / 1024) / (time.time() - t0)
```
Comment (mwaskom) [inline modal/_utils/blob_utils.py:300]: Nit
```suggestion
    throughput_mib_s = size_mib / (time.time() - t0)
```

File: modal/_utils/blob_utils.py [hunk 4]
```diff
@@ -288,11 +289,17 @@ async def _blob_upload(
 
 
 async def blob_upload(payload: bytes, stub: ModalClientModal) -> str:
+    logger.debug(f"Uploading large blob of size {len(payload) / 1024 / 1024:.2f} MiB")
```
Comment (mwaskom) [inline modal/_utils/blob_utils.py:292]: Could compute `size_mib` once first and use here too.

**Review by mwaskom — COMMENTED** [review]
Looks helpful! If you have it handy, could you share what the output looks like for one of your repros? My only concern is that we avoid accidentally making the debug logs uninterpretably-verbose for production-scale Apps.

File: modal/_utils/blob_utils.py (hunk 1 above)
Comment (thundergolfer) [reply modal/_utils/blob_utils.py:334]: Turns out you can get a `0` out of that duration calculation in tests too 😖

**Review by mwaskom — APPROVED** [review]

**Pull request merged**